vnext/include/ReactUWP/Views/KeyboardEventHandler.h

```diff
@@ -0,0 +1,153 @@
+  using namespace Windows::UI::Xaml;
+  using namespace Windows::UI::Xaml::Controls;
+  using namespace Windows::UI::Xaml::Input;
+  using namespace Windows::Foundation;
+  using namespace Windows::UI::Xaml::Media;
+  using namespace Windows::System;
+  using namespace Windows::UI::Core;
+}
+
+using namespace std;
+
+namespace react { namespace uwp {
+  enum class HandledEventPhase
+  {
+    CAPTURING = 1, // match the value with EventPhase in React. EventPhase includes None, Capturing, AtTarget, Bubbling
+    BUBBLING = 3
+  };
+
+  struct ModifiedKeyStatue
+  {
+    bool altKey{ false };
+    bool ctrlKey{ false };
+    bool metaKey{ false };
+    bool shiftKey{ false };
+    bool capLocked{ false };
+  };
+
+  struct ReactKeyboardEvent: ModifiedKeyStatue
+  {
+    int64_t target = 0;
+    string key{};
+  };
+
+  struct KeyboardEvent: ModifiedKeyStatue
+  {
+    HandledEventPhase handledEventPhase{ HandledEventPhase::BUBBLING };
+    string key{};
+  };
+
+  typedef std::function<void(winrt::IInspectable const&, winrt::KeyRoutedEventArgs const&)> KeyboardEventCallback;
+
+  class KeyboardEventBaseHandler {
+  public:
+    KeyboardEventBaseHandler(KeyboardEventCallback&& keyDown, KeyboardEventCallback&& keyUp);
+    virtual ~KeyboardEventBaseHandler() {};
+
+    virtual void hook(XamlView xamlView) = 0;
+    virtual void unhook() = 0;
+
+  protected:
+    KeyboardEventCallback m_keyDownCallback;
+    KeyboardEventCallback m_keyUpCallback;
+  };
+
+  class PreviewKeyboardEventHandler: public KeyboardEventBaseHandler
+  {
+  public:
+    PreviewKeyboardEventHandler(KeyboardEventCallback&& keyDown, KeyboardEventCallback&& keyUp);
+
+    void hook(XamlView xamlView);
+    void unhook();
+
+  private:
+    winrt::UIElement::PreviewKeyDown_revoker m_previewKeyDownRevoker{};
+    winrt::UIElement::PreviewKeyUp_revoker m_previewKeyUpRevoker{};
+  };
+
+  class KeyboardEventHandler : public KeyboardEventBaseHandler
+  {
+  public:
+    KeyboardEventHandler(KeyboardEventCallback&& keyDown, KeyboardEventCallback&& keyUp);
+
+    void hook(XamlView xamlView);
+    void unhook();
+
+  private:
+    winrt::UIElement::KeyDown_revoker m_KeyDownRevoker{};
+    winrt::UIElement::KeyUp_revoker m_KeyUpRevoker{};
+  };
+
+  class PreviewKeyboardEventHandlerOnRoot: public PreviewKeyboardEventHandler
+  {
+  public:
+    PreviewKeyboardEventHandlerOnRoot(const std::weak_ptr<IReactInstance>& reactInstance);
+
+  private:
+    void OnPreKeyDown(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args);
+    void OnPreKeyUp(winrt::IInspectable const& sender, winrt::KeyRoutedEventArgs const& args);
+
+    void DispatchEventToJs(string name, winrt::KeyRoutedEventArgs const& args);
+    std::weak_ptr<IReactInstance> m_wkReactInstance;
+  };
+
+  class HandledKeyboardEventHandler
+  {
+  public:
+    enum class KeyboardEventPhase
+    {
+      PreviewKeyUp,
+      PreviewKeyDown,
+      KeyUp,
+      KeyDown
+    };
+
+    enum class KeyboardType
```

Comment:
KeyboardType doesn't seem like the right name for this. I've seen lots of names for this type of enum. EventType, TransitionType, etc. But KeyboardType sounds like what type of Keyboard :)

Reply:
I found I didn't use it anymore. remove it

---
In reply to: [295559756](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295559756) [](ancestors = 295559756)